docs: Add stake transition documentation (#351)
babylonlabs-io/babylon: docs/stake-transition.md

# Transition of Existing Stakes to the Babylon Chain

This document is intended for Bitcoin stakers who have
submitted confirmed Bitcoin staking transactions
on the Bitcoin blockchain before the launch of
the phase-2 Babylon blockchain (e.g., via participation

> in one of the phase-1 networks).

Reviewer: ```suggestion
in one of the phase-1 caps).
```

Author: Networks refers to the testnet, mainnet, devnet etc. not the caps themselves

> It outlines the eligibility criteria for the registration of

Reviewer: ```suggestion
It outlines the eligibility criteria for the transition of
```

Should we stick with either `transition` or `registration`? They are both used and I feel a bit confused. Or we can add one line to say they are interchangeable

Author: I prefer using registration instead of transition, as transition might imply that the funds are transferred from Bitcoin to somewhere else. Will update transition elsewhere to registration

the stakes, as well as the required data to submit to the
Babylon blockchain to gain voting power and start earning
rewards.

> Note: Each Babylon blockchain network is associated
> with a Bitcoin network (e.g.,
> test environments connect to the BTC signet,
> while production connects to the BTC mainnet).
> For the remainder of the document, we will
> assume that you are registering a stake that
> has been created on the corresponding Bitcoin network
> with the expected parameters (e.g.,
> for phase-2 testnet, registering stakes from phase-1 testnet).

## Registration Eligibility

Bitcoin stakers can hold multiple stakes. The following
flowchart outlines the criteria for each stake's transition.
Note that eligibility is assessed on a
*per-stake* basis, meaning that some stakes
may be eligible for registration while others are not.

![Stake Registration Eligibility](./static/stake-register-eligibility.png)

The following subsections explain the different
conditions presented in the diagram.

> #### Stake Validity

Author: It's on purpose

A stake is valid and eligible for registration on the Babylon blockchain if
it meets the following criteria:
1. **Confirmed on the Bitcoin ledger**:
   The staking transaction must have received sufficient
   Bitcoin confirmations to be considered final.
   The Babylon blockchain defines this in its
   [module parameters](../proto/babylon/btccheckpoint/v1/params.proto)
   (expected to be set to 10 confirmations for testnets and mainnet).
2. **Compliance with Staking Parameters**:
   Staking transactions must adhere to the Bitcoin block-range parameters
   defined by the Babylon blockchain, which are consistent with those of

>    the Phase-1 system (e.g., covenant keys, and minimum and maximum stake requirements).

Reviewer: ```suggestion
   the Phase-1 system (e.g., covenant keys, and minimum and maximum stake requirements).
```
We are using both `Phase-1` and `phase-1`. Should we stick with `Phase-1` and `Phase-2`?

   Valid stakes include:
   1. **Active stakes**: As classified by the phase-1 system.
   2. **Overflow stakes**: Stakes that were valid but did not fit within phase-1
      staking caps will be accepted in phase-2.
   3. **Stakes with malformed OP_RETURN**: Transactions rejected in phase-1
      due to malformed `OP_RETURN` outputs are still valid for the Babylon blockchain
      if they meet all other criteria. Unlike phase-1, the Babylon blockchain
      does not require or check `OP_RETURN` outputs.

#### Stake in the Allow-List

To ensure a secure and controlled onboarding of existing stakes
and the creation of new ones,
the Babylon blockchain employs an initial stake ingestion policy.
This is implemented via an allow-list that specifies which staking
transactions can be registered by their transaction hashes.

It has two important properties:
* **Hardcoded Allow-List**: This list is embedded in the Babylon blockchain
  codebase and is visible to all users.

> * **Automatic Expiration**: The allow-list functionality has a fixed expiration
>   date to prevent abuse.

Reviewer: ```suggestion
* **Automatic Expiration**: The allow-list functionality has a fixed expiration
  Babylon block height to prevent abuse.
```
or just block height

> The allow-list contents and its expiration date will be chosen

Reviewer: ```suggestion
The allow-list contents and its expiration block height will be chosen
```
